sql-plugin/src/main/scala/com/nvidia/spark/rapids/ScalableTaskCompletion.scala

```diff
@@ -0,0 +1,257 @@
+  // grabbed should always be
+  //
+  // ScalableTaskCompletion -> TopLevelTaskCompletion -> UserTaskCompletion
+  //
+  // this is to avoid any potential deadlocks
+
+  /**
+   * Wrapper around a user callback function.
+   */
+  private class UserTaskCompletion(val tc: TaskContext,
+      f: Either[() => Unit, TaskContext => Unit])
+      extends TaskCompletionCallbackHandle
+          with Function[TaskContext, Unit] {
+    private var wasCalled = false
+
+    override def removeCallback(): Unit = {
+      val topLevel = ScalableTaskCompletion.getTopLevel(tc.taskAttemptId())
+      if (topLevel != null) {
+        topLevel.remove(this)
+      }
+    }
+
+    override def apply(tc: TaskContext): Unit = synchronized {
+      if (!wasCalled) {
+        wasCalled = true
+        f match {
+          case Left(l) => l()
+          case Right(r) => r(tc)
+        }
+      }
+    }
+
+    override def removeAndCall(): Unit = {
+      removeCallback()
+      apply(tc)
+    }
+  }
+
+  /**
+   * Testing wrapper around the user callback function. It is not common and is here to allow
+   * us to still have things behave correctly in some unit tests.
+   */
+  private class TestTaskCompletion(f: () => Unit)
+      extends TaskCompletionCallbackHandle {
+    override def removeCallback(): Unit = {
+      // NOOP
+    }
+
+    override def removeAndCall(): Unit = f()
+  }
+
+  /**
+   * Keeps track of user callbacks for a given task. It is also the callback from
+   * Spark itself.
+   */
+  private class TopLevelTaskCompletion extends Function[TaskContext, Unit] {
+    private val pending = new util.HashSet[UserTaskCompletion]()
+    private var callbacksDone = false
+
+    override def apply(tc: TaskContext): Unit = {
+      var error: Throwable = null
+      synchronized {
+        pending.forEach { utc =>
+          try {
+            utc(tc)
+          } catch {
+            case t: Throwable =>
+              if (error == null) {
+                error = t
+              } else {
+                error.addSuppressed(t)
+              }
+          }
+        }
+        pending.clear()
+        callbacksDone = true
+      }
+      try {
+        ScalableTaskCompletion.removeTopLevel(tc.taskAttemptId())
+      } catch {
+        case t: Throwable =>
+          if (error == null) {
+            error = t
+          } else {
+            error.addSuppressed(t)
+          }
+      }
+      if (error != null) {
+        throw error
+      }
+    }
+
+    def add(u: UserTaskCompletion): Unit = synchronized {
+      if (callbacksDone) {
+        // Added a callback after it was done calling them back already
+        u(u.tc)
+      } else {
+        pending.add(u)
+      }
+    }
+
+    def remove(u: UserTaskCompletion): Unit = synchronized {
+      pending.remove(u)
+    }
+
+    def removeAllAndShutdown(): Unit = synchronized {
+      pending.forEach { utc =>
+        utc(utc.tc)
```

Comment:
Do we need the same "catch but keep calling all of them before throwing" logic here?  Wondering if the code to call the methods and set callbacksDone should be moved into a private method that can be reused here.

Reply:
We could. This is here only for testing so I didn't add it, but it is simple enough to do.